File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        
        Arrays.sort(nums);
        
        for(int i = 0; i<nums.length; i++){
            while(i > 0 && i < nums.length && nums[i] == nums[i-1]){
                i++;
            }
            int j = i+1;
            int k = nums.length-1;
            while(j < k){
                
                int sum = nums[i]+nums[j]+nums[k];
                if(sum == 0){
                    List<Integer> triplet = new ArrayList<>();
                    triplet.add(nums[i]);
                    triplet.add(nums[j]);
                    triplet.add(nums[k]);

                    ans.add(triplet);
                    j++;
                    k--;
                    while(j < nums.length-1 && nums[j] == nums[j-1]){
                        j++;
                    }
                    while(k > 0 && nums[k] == nums[k+1]){
                        k--;
                    }
                    
                } else if(sum < 0){
                    j++;
                } else {
                    k--;
                }
            }
        }
        
        return ans;
    }
}

File: 15-3sum/15-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let res = [];
    let n = nums.length;
    if (n < 3) {
        return res;
    }
    nums.sort((a, b) => a - b);
    for (let i = 0; i < n - 2; i++) {
        if (nums[i] > 0) {
            break;
        }
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        let l = i + 1;
        let r = n - 1;
        while (l < r) {
            let sum = nums[i] + nums[l] + nums[r];
            if (sum === 0) {
                res.push([nums[i], nums[l], nums[r]]);
                while (l < r && nums[l] === nums[l + 1]) {
                    l++;
                }
                while (l < r && nums[r] === nums[r - 1]) {
                    r--;
                }
                l++;
                r--;
            } else if (sum < 0) {
                l++;
            } else {
                r--;
            }
        }
    }
    return res;
};